### ISSUE TITLE
Pharo image freeze on Windows after some hours Idle

### ISSUE THREAD

**Issue by VincentBlondeau** [guillep/pharo-core#8170]
**Describe the bug**
After some time keeping open a pharo image with the headless image, the image becomes non-responsive but is still open.

The error is:
Error: Bad BitBlt arg (Fraction?); proceed to convert.
[dump.txt](https://github.com/pharo-project/pharo/files/5739298/dump.txt)

**To Reproduce**
1. Open an image
2. Wait some time
3. See that it is frozen

**Expected behavior**
No freeze

**Version information:**
VM: Win32 - x86_64 - Win64 - CoInterpreterWithQueueFFI VMMaker-tonel.1 uuid: be795315-f06d-0d00-aab5-813a0e395221 Oct 13 2020
StackToRegisterMappingCogit VMMaker-tonel.1 uuid: be795315-f06d-0d00-aab5-813a0e395221 Oct 13 2020
e829a1da - Commit: e829a1da - Date: 2020-10-13 15:43:09 +0200

Image: Pharo9.0.0 [Build information: Pharo-9.0.0+build.997.sha.5bf82e9abc5aea283f18e74bbe27ee9464483ea9 (64 Bit)]

**Expected development cost**
No clue

**Additional context**
It may be due also to a windows session locking

**Comment by VincentBlondeau** [comment]
I ran the WinDbg utility on the frozen image for each running thread. Here is the the result:
[WinDbg-SDLCrash.txt](https://github.com/pharo-project/pharo/files/5763965/WinDbg-SDLCrash.txt)

I try to let the image still in a freeze mode, let me know if other commands should be run.

**Comment by tesonep** [comment]
Thanks for the debug info @VincentBlondeau 
I will check why we are not getting SDL events.

**Comment by VincentBlondeau** [comment]
I can reproduce the issue when the Windows user session got locked and when I login in, the image freeze.
Let me know if I can help you with more debugging info.

NB: this issue will be critical to be fixed for Lifeware.

**Comment by Ducasse** [comment]
We will fix it :)

**Comment by fxgallego** [comment]
I can't reproduce the same with Pharo 9 but something similar happens when you unplug one display if you have more than one connected. P9 window is still visible but hang and doesn't redraw anything. In my case it doesn't depends on which display the Pharo windows is, it alway hangs when you unplug one. Interesting that the Pharo Launcher doesn't shows the same behavior.
Having the image open for days, locking, unlocking windows session, even suspend works ok, but unplug display did the trick.

System Reporter says:
```
Image
-----
C:\Users\scalvo\Documents\Pharo\images\Pharo 9.0 - 64bit (development version, latest)\Pharo 9.0 - 64bit (development version, latest).image
Pharo9.0.0
Build information: Pharo-9.0.0+build.1102.sha.2b9380cc3a7f82211829e244089f0f04399398d4 (64 Bit)
Unnamed

Virtual Machine
---------------
C:\Users\scalvo\Documents\Pharo\vms\90-x64\Pharo.exe
CoInterpreterWithQueueFFI VMMaker-tonel.1 uuid: be795315-f06d-0d00-aab5-813a0e395221 Oct 13 2020
StackToRegisterMappingCogit VMMaker-tonel.1 uuid: be795315-f06d-0d00-aab5-813a0e395221 Oct 13 2020
e829a1da - Commit: e829a1da - Date: 2020-10-13 15:43:09 +0200

```

**Comment by VincentBlondeau** [comment]
It is interesting. I have the issue when the Remote Desktop Session to the system running Pharo is over or locked.
It maybe linked to the screen addition then

**Comment by EwanDawson** [comment]
I experienced the same issue today, with Pharo 9 running on WIndows 10 21H1. I've attached my debug log.
[PharoDebug.log](https://github.com/pharo-project/pharo/files/6906047/PharoDebug.log)

**Comment by Ducasse** [comment]
Thanks a lot for the report.

**Comment by VincentBlondeau** [comment]
Actually, the let the image in idle seems fine. It survived the night. However, I succeeded to crash it by connecting through Remote Windows Desktop to the same session where it was open. Thanks to https://github.com/pharo-project/opensmalltalk-vm/issues/125 I can have the stack:
[pharo-28556.log](https://github.com/pharo-project/pharo/files/7336335/pharo-28556.log)

It seems to be a deadlock to me, this stack is present twice:

Process         0x93438ce0 priority 40
          0x6a7fe8 M [] in Semaphore>critical: 0x9343fce0: a(n) Semaphore
          0x6a8028 M FullBlockClosure(BlockClosure)>ensure: 0x9a7b4bc0: a(n) FullBlockClosure
          0x6a8068 M Semaphore>critical: 0x9343fce0: a(n) Semaphore
          0x6a80b8 I Mutex>critical: 0x9343fbc0: a(n) Mutex
          0x6a8108 I OSSDL2FormRenderer>updateAreas:immediate: 0x9343bb78: a(n) OSSDL2FormRenderer
          0x6a8158 I OSWorldRenderer>updateDamage: 0x93401d18: a(n) OSWorldRenderer

**Comment by tesonep** [comment]
Can you share the log file?
Thanks a lot.

**Comment by VincentBlondeau** [comment]
Here it is:
[pharo_log.txt](https://github.com/pharo-project/pharo/files/7336586/pharo_log.txt)

**Comment by tesonep** [comment]
I am checking it, but it looks related with #10156
It seems if an error occurs inside the Mutex, as the ensure is not correctly called if there is an error (and a debugger is openned) the mutex is never release, so next updates never occurs 😮

**Comment by tesonep** [comment]
Thanks a lot @VincentBlondeau for the information

**Comment by guillep** [comment]
Just for documentation purposes, I think @tesonep got an easier and 100% reliable way to reproduce it in seconds:

 - open Pharo
 - log out session (letting pharo open)
 - log in again

And it is frozen :)

**Comment by demarey** [comment]
fixed in #10179 and #10188
